ocaml/xapi/eventgen.ml

```diff
@@ -11,248 +11,200 @@
+    Dm_api.relations_of_api api
+    |> List.concat_map (fun (a, b) -> [(a, b); (b, a)])
+  in
+  (* Find an object related to the input field using the datamodel. *)
+  let find_related_object = function
+    | DT.{field_name; ty= Ref _; _} ->
+        let this_end = (obj_name, field_name) in
+        List.assoc_opt this_end symmetric
+        |> Option.map (fun (other_object, _) -> (other_object, field_name))
+    | _ ->
+        None
+  in
+  let fields = Datamodel_utils.fields_of_obj obj in
+  List.filter_map find_related_object fields
 
-let obj_references_table : (string, (string * string) list) Hashtbl.t =
-  Hashtbl.create 30
+(* For each object, precompute a list of related objects [(object,
+   field); ...] and store it in a hash table.
 
-(* populate obj references table *)
-let _ =
-  List.iter
-    (fun obj ->
-      let obj_name = obj.Datamodel_types.name in
-      Hashtbl.replace obj_references_table obj_name
-        (compute_object_references_to_follow obj_name)
-    )
-    (Dm_api.objects_of_api Datamodel.all_api)
+   If looking up an entry for some object, "foo", yields a list
+   containing ("bar", "baz"), then it must be the case that "foo"'s
+   field "baz" (which is necessarily of type Ref _) is related to some
+   field within "bar". In which case, the database will have already
+   updated the related field(s) and we must emit events for those fields. *)
+let obj_references_table : (string, (string * string) list) Hashtbl.t =
+  let table = Hashtbl.create 64 in
+  let populate_follows (obj : Datamodel_types.obj) =
+    let follows = compute_object_references_to_follow obj.name in
+    Hashtbl.replace table obj.name follows
+  in
+  Dm_api.objects_of_api Datamodel.all_api |> List.iter populate_follows ;
+  table
 
 let follow_references (obj_name : string) =
   Hashtbl.find obj_references_table obj_name
 
-(** Compute a set of modify events but skip any for objects which were missing
-    (must have been dangling references) *)
-let events_of_other_tbl_refs other_tbl_refs =
-  List.concat_map
-    (fun (tbl, fld, x) ->
-      try [(tbl, fld, x ())]
-      with _ ->
-        (* Probably means the reference was dangling *)
-        warn "skipping event for dangling reference %s: %s" tbl fld ;
-        []
-    )
-    other_tbl_refs
+(* Compute a modify event's snapshot by attemping to invoke its
+   getter. If the record cannot be found, the event is dropped (as the
+   reference is probably dangling). *)
+let snapshots_of_other_tbl_refs other_tbl_refs =
+  let try_get_records (table, field, getter) =
+    try Some (table, field, getter ())
+    with _ ->
+      (* Probably means the reference was dangling *)
+      warn "skipping event for dangling reference %s: %s" table field ;
+      None
+  in
+  List.filter_map try_get_records other_tbl_refs
 
 open Xapi_database.Db_cache_types
 open Xapi_database.Db_action_helper
 
-let database_callback_inner event db context =
+let is_valid_ref db = function
```

Comment:
I assume the object that is being dropped is just a pointer - hence it is not expensive to make this call because it does not allocate.

Reply:
>I believe the is_valid_ref mechanism is approximate and any unintended side effect is benign.

I think that, as the DB only stores things of type `String`, `Set` and `Pairs` (`Schema.Value.t`), the underlying datamodel types are not visible at this level. So any `String` may or may not be a reference, and if so the reference may or may not be valid (anymore). So a field like `name_label`, which isn't a ref, my contain a string that happens to be valid ref. But then it gets filtered out in `compute_other_table_events`/`follow_references`, because it won't be a defined relation? If not, indeed it looks benign.